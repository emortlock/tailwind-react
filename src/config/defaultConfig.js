export default {
  radius: 'rounded',
  spacing: {
    zero: 0,
    sm: 2,
    md: 4,
    lg: 8,
  },
  container: {
    sm: 'xl',
    md: '3xl',
    lg: '4xl',
  },
  text: {
    size: {
      body: ['sm', 'base', 'lg'],
      title: ['lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
    },
    family: {
      body: 'sans',
      subtitle: 'sans',
      title: 'sans',
    },
  },
  baseColors: {
    primaryLight: 'blue-light',
    primary: 'blue',
    primaryDark: 'blue-dark',
    secondaryLight: 'grey',
    secondary: 'grey-dark',
    secondaryDark: 'grey-darker',
    successLight: 'green-light',
    success: 'green',
    successDark: 'green-dark',
    dangerLight: 'red-light',
    danger: 'red',
    dangerDark: 'red-dark',
    warningLight: 'orange-lighter',
    warning: 'orange-light',
    warningDark: 'orange',
    infoLight: 'teal-lighter',
    info: 'teal-light',
    infoDark: 'teal',
  },
  textColors: {
    body: 'grey-darkest',
    link: 'blue-dark',
    linkDark: 'blue-darker',
    emphasis: 'black',
    on: {
      primaryLight: 'white',
      primary: 'white',
      primaryDark: 'white',
      secondaryLight: 'white',
      secondary: 'white',
      secondaryDark: 'white',
      successLight: 'white',
      success: 'white',
      successDark: 'white',
      dangerLight: 'white',
      danger: 'white',
      dangerDark: 'white',
      warningLight: 'black',
      warning: 'black',
      warningDark: 'black',
      infoLight: 'black',
      info: 'black',
      infoDark: 'black',
    },
  },
}
