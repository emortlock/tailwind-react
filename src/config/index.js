export { default as defaultConfig } from './defaultConfig'
export { default as TailwindConfig } from './TailwindConfig'
export { default as withConfig } from './withConfig'
