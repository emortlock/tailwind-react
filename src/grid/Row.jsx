import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const Row = ({ is, children, nowrap, gutter, theme, ...rest }) => {
  const gutterSpacing =
    gutter && (gutter === true ? theme.spacing.md : theme.spacing[gutter])

  return (
    <BaseComponent
      is={is}
      flex={nowrap || [true, 'wrap']}
      nm={
        gutter
          ? { l: gutterSpacing, b: !nowrap ? gutterSpacing : undefined }
          : undefined
      }
      className="list-reset"
      {...rest}
    >
      {gutter
        ? React.Children.map(children, child =>
            React.cloneElement(child, {
              p: { l: gutterSpacing },
              m: { b: gutterSpacing },
            }),
          )
        : children}
    </BaseComponent>
  )
}

Row.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  nowrap: PropTypes.bool,
  gutter: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
}

Row.defaultProps = {
  is: 'ul',
  children: undefined,
  nowrap: false,
  gutter: false,
}

export { Row as component }
export default withTheme(Row)
