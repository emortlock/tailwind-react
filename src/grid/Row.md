Follows the component behaviour described on the Tailwind CSS [grid component page](https://tailwindcss.com/docs/examples/grids).

Example:

```jsx
<Row>
  <Col>
    <div className="bg-grey p-4 text-center">1/1</div>
  </Col>
  <Col size="1/2">
    <div className="bg-grey-light p-4 text-center">1/1</div>
  </Col>
  <Col size="1/2">
    <div className="bg-grey p-4 text-center">1/2</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey-light p-4 text-center">1/3</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey p-4 text-center">1/3</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey-light p-4 text-center">1/3</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey-light p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey-light p-4 text-center">1/4</div>
  </Col>
</Row>
```

Gutters:

```jsx
<Row gutter>
  <Col>
    <div className="bg-grey p-4 text-center">1/1</div>
  </Col>
  <Col size="1/2">
    <div className="bg-grey-light p-4 text-center">1/1</div>
  </Col>
  <Col size="1/2">
    <div className="bg-grey p-4 text-center">1/2</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey-light p-4 text-center">1/3</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey p-4 text-center">1/3</div>
  </Col>
  <Col size="1/3">
    <div className="bg-grey-light p-4 text-center">1/3</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey-light p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey p-4 text-center">1/4</div>
  </Col>
  <Col size="1/4">
    <div className="bg-grey-light p-4 text-center">1/4</div>
  </Col>
</Row>
```

Responsive column widths:

```jsx
<Row gutter>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey p-4 h-12" />
  </Col>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey-light p-4 h-12" />
  </Col>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey p-4 h-12" />
  </Col>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey-light p-4 h-12" />
  </Col>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey p-4 h-12" />
  </Col>
  <Col size={{ def: 'full', sm: '1/2', md: '1/3', lg: '1/4', xl: '1/6' }}>
    <div className="bg-grey-light p-4 h-12" />
  </Col>
</Row>
```

Auto column widths:

```jsx
<Row nowrap>
  <Col size="auto">
    <div className="bg-grey p-4 h-12" />
  </Col>
  <Col size="auto">
    <div className="bg-grey-light p-4 h-12" />
  </Col>
  <Col size="auto">
    <div className="bg-grey p-4 h-12" />
  </Col>
</Row>
```
