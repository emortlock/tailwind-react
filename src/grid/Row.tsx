import type { FlexProps } from 'tailwind-react-primitives'
import type { Theme, WithTheme } from '../theme'

import React from 'react'

import { RawFlex as Flex } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface RowProps<E extends HTMLElement = HTMLUListElement>
  extends WithTheme,
    FlexProps<E> {
  nowrap?: boolean
  gutter?: boolean | keyof Theme['spacing']
}

export const RawRow = <E extends HTMLElement = HTMLUListElement>({
  as = 'ul',
  children,
  nowrap,
  gutter,
  theme,
  ...rest
}: RowProps<E>) => {
  const gutterSpacing =
    typeof gutter === 'string'
      ? theme.spacing[gutter]
      : gutter && theme.spacing.md

  return (
    <Flex
      as={as}
      wrap={!nowrap}
      nm={
        gutter
          ? {
              l: gutterSpacing || undefined,
              b: !nowrap ? gutterSpacing || undefined : undefined,
            }
          : undefined
      }
      {...rest}
    >
      {gutter
        ? React.Children.map(children, (child) =>
            React.isValidElement(child)
              ? React.cloneElement(child, {
                  p: { l: gutterSpacing },
                  m: { b: gutterSpacing },
                })
              : child,
          )
        : children}
    </Flex>
  )
}

export const Row = withTheme(RawRow)
