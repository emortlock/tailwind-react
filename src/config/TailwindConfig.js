import React from 'react'
import defaultConfig from './defaultConfig'

export default React.createContext(defaultConfig)
