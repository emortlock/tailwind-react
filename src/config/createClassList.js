import merge from 'lodash.merge'
import defaultConfig from './defaultConfig'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectConfig = {}) => {
  const config = merge(defaultConfig, projectConfig)

  return [
    config.radius,
    ...reduceMultiplePrefixes(config.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(config.container, 'max-w'),
    ...mapArray(config.text.size.body, 'text'),
    ...mapArray(config.text.size.title, 'text'),
    ...mapObject(config.text.family, 'font'),
    ...reduceMultiplePrefixes(config.baseColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(config.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
