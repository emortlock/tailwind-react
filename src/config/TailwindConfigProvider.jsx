import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultConfig from './defaultConfig'
import TailwindConfig from './TailwindConfig'

const TailwindConfigProvider = ({ config, children }) => {
  const mergedConfig = merge(defaultConfig, config)

  return (
    <TailwindConfig.Provider value={mergedConfig}>
      {children}
    </TailwindConfig.Provider>
  )
}

TailwindConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindConfigProvider.defaultProps = {
  config: {},
  children: undefined,
}

export default TailwindConfigProvider
