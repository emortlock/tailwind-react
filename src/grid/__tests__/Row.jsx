import React from 'react'
import { shallow } from 'enzyme'

import { component as Row } from '../Row'
import { defaultConfig } from '../../config'

const setup = (testProps = {}) => {
  const props = Object.assign({ config: defaultConfig }, testProps)

  const wrapper = shallow(<Row {...props} />)

  return {
    props,
    wrapper,
  }
}

describe('Row', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ gutter: true })

    expect(wrapper).toMatchSnapshot()
  })
})
