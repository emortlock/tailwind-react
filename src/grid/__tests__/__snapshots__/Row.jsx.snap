// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders matching snapshot 1`] = `
<BaseComponent
  flex={
    Array [
      true,
      "wrap",
    ]
  }
  is="div"
  nm={
    Object {
      "b": 4,
      "l": 4,
    }
  }
/>
`;
